### ISSUE TITLE
LTO build fails on ARM64

### ISSUE THREAD

**Issue by MagaTailor** [freeorion/freeorion#1238]
Another problem with LTO and freeorion on an ARM platform - as Ian Fleming would have it, twice is coincidence?

Unlike on [ARM](https://github.com/freeorion/freeorion/issues/1064), aarch64 fails to link:

```shell
[ 61%] Linking CXX executable ../freeoriond
/tmp/ccgVXBvE.ltrans21.ltrans.o: In function `Fighter::~Fighter()':
<artificial>:(.text+0x6d4): undefined reference to `vtable for Fighter'
<artificial>:(.text+0x6d8): undefined reference to `vtable for Fighter'
<artificial>:(.text+0x6f4): undefined reference to `VTT for Fighter'
<artificial>:(.text+0x6f8): undefined reference to `VTT for Fighter'
/tmp/ccgVXBvE.ltrans21.ltrans.o: In function `Fighter::~Fighter()':
<artificial>:(.text+0x74c): undefined reference to `vtable for Fighter'
<artificial>:(.text+0x750): undefined reference to `vtable for Fighter'
<artificial>:(.text+0x76c): undefined reference to `VTT for Fighter'
<artificial>:(.text+0x770): undefined reference to `VTT for Fighter'
/tmp/ccgVXBvE.ltrans4.ltrans.o: In function `boost::detail::sp_counted_impl_p<Fighter>::dispose()':
<artificial>:(.text+0xe30): undefined reference to `vtable for Fighter'
<artificial>:(.text+0xe34): undefined reference to `vtable for Fighter'
<artificial>:(.text+0xe48): undefined reference to `VTT for Fighter'
<artificial>:(.text+0xe4c): undefined reference to `VTT for Fighter'
/tmp/ccgVXBvE.ltrans8.ltrans.o: In function `Field::~Field()':
<artificial>:(.text+0x6ec): undefined reference to `vtable for Field'
<artificial>:(.text+0x6f0): undefined reference to `vtable for Field'
<artificial>:(.text+0x70c): undefined reference to `VTT for Field'
<artificial>:(.text+0x710): undefined reference to `VTT for Field'
/tmp/ccgVXBvE.ltrans8.ltrans.o: In function `Field::~Field()':
<artificial>:(.text+0x764): undefined reference to `vtable for Field'
<artificial>:(.text+0x768): undefined reference to `vtable for Field'
<artificial>:(.text+0x784): undefined reference to `VTT for Field'
<artificial>:(.text+0x788): undefined reference to `VTT for Field'
/tmp/ccgVXBvE.ltrans20.ltrans.o: In function `boost::python::detail::caller_arity<2u>::impl<bool (*)(Field const&, UniverseObject const&), boost::python::default_call_policies, boost::mpl::vector3<bool, Field const&, UniverseObject const&> >::operator()(_object*, _object*) [clone .constprop.547]':
<artificial>:(.text+0x848): undefined reference to `vtable for Field'
<artificial>:(.text+0x84c): undefined reference to `vtable for Field'
<artificial>:(.text+0x860): undefined reference to `VTT for Field'
<artificial>:(.text+0x864): undefined reference to `VTT for Field'
collect2: error: ld returned 1 exit status
server/CMakeFiles/freeoriond.dir/build.make:473: recipe for target 'freeoriond' failed
````

The system is Ubuntu 16.04, `gcc version 5.4.0 20160609` and `boost 1.58`. 

Any ideas what could be causing this? `libstdc++` was definitely compiled with the same gcc version, using the default, new cxx abi.

**Comment by o01eg** [comment]
I suppose destructors of `Field` and `Fighter` should be virtual. Their superclass `UniverseObject` have virtual destructor.

**Comment by o01eg** [comment]
Also, I tried build lto release on x86_64 but get segfault before main.

**Comment by geoffthemedio** [comment]
> I would suppose destructors of `Field` and `Fighter` should be virtual.

Why not the same for `System`, `Planet`, etc. then?

**Comment by o01eg** [comment]
I don't know but `UniverseObject` have virtual destructor. Maybe somewhere destructors are called by poitnter to base class only for `Field` and `Fighter`

**Comment by geoffthemedio** [comment]
Overriding a base class' virtual function makes the derived class' function also virtual. http://stackoverflow.com/questions/2198379/are-virtual-destructors-inherited

**Comment by MagaTailor** [comment]
@geoffthemedio If a mainstream platform like `x86_64` is crashing too, as per @o01eg's [comment](https://github.com/freeorion/freeorion/issues/1238#issuecomment-274607542),  maybe a new issue should be opened? The other platforms do manage to finish building.

**Comment by geoffthemedio** [comment]
Sure, that sounds like a separate issue.

**Comment by o01eg** [comment]
I cann't be sure if it bug in the freeorion not compiler. I'm still on gcc-4.

**Comment by MagaTailor** [comment]
Slightly more verbose [info](https://gist.github.com/MagaTailor/728d7237098c64e4292ed0e5b05c394a) from a debug LTO build.

**Comment by geoffthemedio** [comment]
Try removing the `#if BOOST_VERSION >= 106100` around the ~Field and ~Fighter definitions. (Or adding an alternative if case that would apply just in whatever build environment you're using...)

**Comment by MagaTailor** [comment]
You mean it should also trigger for my boost version?

**Comment by geoffthemedio** [comment]
or whatever "LTO" or "aarch64" are, if there's a test for them with a preprocessor define.

I assume this problem is not related to boost at all, so testing the boost version is irrelevant.

**Comment by MagaTailor** [comment]
I tried removing that define everywhere it was present but it didn't help apparently.

**Comment by geoffthemedio** [comment]
So you had just, eg:
````
public:
    ~System() {}
protected:
````
instead of
````
#if BOOST_VERSION >= 106100
public:
#endif
    ~System() {}
#if BOOST_VERSION >= 106100
protected:
#endif
````

**Comment by MagaTailor** [comment]
Precisely.

**Comment by geoffthemedio** [comment]
Change to `~Field();` in the header, and then add to Field.cpp:
````
Field::~Field()
{}
````
The discussion here
http://stackoverflow.com/questions/3065154/undefined-reference-to-vtable
has a comment that says that moving the definition of the destructor into the definition file got rid of the initial error and revealed the actual problem.

**Comment by MagaTailor** [comment]
Ok, so I did that as well as edited `Fighter.h`/`cpp` and the build completed successfully. Cool!

**Comment by MagaTailor** [comment]
Could be unrelated to the fix but #1064 doesn't crash any more (gcc 4.9).

**Comment by o01eg** [comment]
I'll check if it also fixes crash on x86_64 with LTO compilation.

**Comment by MagaTailor** [comment]
Well, it seemed implausible that the patch had been responsible for fixing the other issue and it didn't :)

It was a compiler bug all along, as freeorion doesn't crash after dumping `-mfpu=neon -ftree-vectorize` in favour of `-mfpu=vfpv4` instead. The more aggressive flags work fine for non-LTO builds.

**Comment by Vezzra** [comment]
> It was a compiler bug all along...

@MagaTailor, do I understand correctly, changing the flags solved #1064 for you? Just want to know, because in this case we need to relabel #1064 from wontfix to invalid (not that important, but want to be thorough 😉).

**Comment by MagaTailor** [comment]
Yes, it looks like I've managed to hit the right combination of flags this time (having tried a few different ones at the time of the original report).
